compiler/rustc_hir_typeck/src/lib.rs

```diff
@@ -414,10 +414,12 @@ enum TupleArgumentsFlag {
 
 fn fatally_break_rust(tcx: TyCtxt<'_>) {
     let handler = tcx.sess.diagnostic();
-    handler.span_bug_no_panic(
-        MultiSpan::new(),
+    // We normally use `handler.bug()` for bugs, but this is an exceptional case, so we do this
+    // instead to avoid an abort.
+    handler.emit_diagnostic(&mut Diagnostic::new(
+        Level::Bug,
         "It looks like you're trying to break rust; would you like some ICE?",
-    );
+    ));
     handler.note("the compiler expectedly panicked. this is a feature.");
     handler.note(
```

Comment:
Why is this not making a `DiagnosticBuilder<'_, Bug>` and appending the `note`s onto that? 🤔

Reply:
Good question. A couple of reasons, AIUI.

First, the API around `Bug` is currently a bit of a mess:
- `Handler::struct_bug` -> `DB<!>`
- `Handler::span_bug` -> `!`
- `Handler::bug` -> `!`
- `Handler::create_bug` -> `DB<Bug>`    (unused)
- `Handler::emit_bug` -> `Bug` (unused)
- `Bug::diagnostic_builder_emit_producing_guarantee` -> `!` (not `Bug`)

I.e. it's a mix of return types, unlike all the other diagnostic levels. I want to straighten this out, but I haven't gotten to it yet. I think part of the problem is that the `EmissionGuarantee` implemented for `!` is tied to `Level::Fatal` rather than `Level::Bug`, even though `Level::Bug` should also cause immediate aborts.

Except, this is the one ICE that doesn't actually abort. So if we used `struct_bug` + `emit` it would abort. I guess we could use `create_bug` + `emit_bug`, but that would require a new type. Maybe good for a follow-up. Or maybe it should just abort? That would make life easier in the long run.

Also, generating distinct notes gives output like this:
```
error: internal compiler error: It looks like you're trying to break rust; would you like some ICE?

note: the compiler expectedly panicked. this is a feature.

note: we would appreciate a joke overview: https://github.com/rust-lang/rust/issues/43162#issuecomment-320764675

note: rustc 1.76.0-dev running on x86_64-unknown-linux-gnu
```
Generating a single (non-fatal) abort with appended notes looks like this:
```
error: internal compiler error: It looks like you're trying to break rust; would you like some ICE?
  |
  = note: the compiler expectedly panicked. this is a feature.
  = note: we would appreciate a joke overview: https://github.com/rust-lang/rust/issues/43162#issuecomment-320764675
  = note: rustc 1.76.0-dev running on x86_64-unknown-linux-gnu
```
It's possible that the latter output is better, but I didn't want to change it. (I have been wondering what is the use of standalone notes, as opposed to notes attached to errors/warnings/etc...)

Reply:
As per our discussion on Zulip: I changed this to a normal aborting ICE, which meant I could just use `struct_bug` and `emit`. It did require changing the notes from standalone diagnostics to notes attached to the bug diagnostic, but that seems fine.

Reply:
Turns out my introduction of `struct_bug` broke some things. So in the end I removed that, and also the removal of `span_bug_no_panic`, and also the changes to this ICE. They can be done in a follow-up once I figure out how to fix `struct_bug` properly.